test/test_torch.py

```diff
@@ -5902,80 +5908,67 @@ def run(device, data, model, optimizer, scaler, loss_fn, skip_iter, try_scaling_
                         optimizer.step()
             return scaler
 
-        # NOTE(mkozuki): With current way of testing, `torch.optim.Adam` is failing in spite of `foreach` and `fused`.
-        #   Giving some flexibility to this test might help.
-        context = contextlib.nullcontext
-        if optimizer_ctor in (torch.optim.Adam, torch.optim.AdamW):
-            from functools import partial
-            context = partial(self.assertRaises, AssertionError)
-        with context():
-            # sets atol=1e-3 because we're comparing pure fp32 arithmetic vs a mixture of fp16 and fp32
-            self._run_scaling_case(
-                device, run, unskipped=3, skipped=1, atol=atol, optimizer_ctor=optimizer_ctor, optimizer_kwargs=optimizer_kwargs,
-            )
-            # this will be picked up by try_pickle within run():
-            try_pickle = True
-            self._run_scaling_case(
-                device, run, unskipped=3, skipped=1, atol=atol, optimizer_ctor=optimizer_ctor, optimizer_kwargs=optimizer_kwargs,
-            )
-
-    @onlyNativeDeviceTypes
-    def test_grad_scaling_autocast(self, device):
-        device = torch.device(device)
-        for optimizer_ctor in (torch.optim.SGD, torch.optim.Adam, torch.optim.AdamW):
-            self._grad_scaling_autocast_test(device=device.type, optimizer_ctor=optimizer_ctor)
-
-    @onlyNativeDeviceTypes
-    def test_grad_scaling_autocast_foreach(self, device):
-        device = torch.device(device)
-        for optimizer_ctor in (torch.optim.SGD, torch.optim.Adam, torch.optim.AdamW):
-            self._grad_scaling_autocast_test(device=device.type, optimizer_ctor=optimizer_ctor, optimizer_kwargs={"foreach": True})
-
-    @onlyNativeDeviceTypes
-    def test_grad_scaling_autocast_fused(self, device):
-        device = torch.device(device)
-        for optimizer_ctor in (torch.optim.Adam, torch.optim.AdamW):
-            self._grad_scaling_autocast_test(device=device.type, optimizer_ctor=optimizer_ctor, optimizer_kwargs={"fused": True})
+        optimizer_ctor = optim_info.optim_cls
+        all_optim_inputs = _get_optim_inputs_including_global_cliquey_kwargs(
+            device, dtype, optim_info, skip=("differentiable"))
+
+        # Compares no scaling + no autocasting against scaling + autocasting.
+        for optim_input in all_optim_inputs:
+
+            # NOTE(mkozuki): With current way of testing, `torch.optim.Adam` is failing in spite of `foreach` and `fused`.
+            #   Giving some flexibility to this test might help.
+            context = contextlib.nullcontext
+            if optimizer_ctor in (torch.optim.Adam, torch.optim.AdamW):
+                from functools import partial
+                context = partial(self.assertRaises, AssertionError)
```

Comment:
It looks like one of the errors is no longer getting raised...which sounds like a good thing! 

It'd be good to know which configs do not pass in the failing SGD tests.

Reply:
When I remove the `if` block, ie: keep `context = contextlib.nullcontext` for all cases of `optimizer_ctor`, I get the following errors for Adam and AdamW optims
```
AssertionError: Tensor-likes are not close!

Mismatched elements: 64 / 64 (100.0%)
Greatest absolute difference: 5.1396894454956055 at index (6, 6) (up to 0.001 allowed)
Greatest relative difference: 10.834877967834473 at index (6, 6) (up to 1e-05 allowed)

To execute this test, run the following from the base repo dir:
     python test/test_torch.py -k test_grad_scaling_autocast_AdamW_cpu_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
```
AssertionError: Tensor-likes are not close!

Mismatched elements: 64 / 64 (100.0%)
Greatest absolute difference: 5.523576259613037 at index (6, 6) (up to 0.001 allowed)
Greatest relative difference: 28.40498161315918 at index (6, 6) (up to 1e-05 allowed)

To execute this test, run the following from the base repo dir:
     python test/test_torch.py -k test_grad_scaling_autocast_Adam_cpu_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

Reply:
The SGD optimizer test is failing for `optim_input.kwargs == {'weight_decay': 0.1, 'maximize': True, 'fused': True}`.

Reply:
The absolute difference of 5 is pretty major and signifies something seems wrong. 

> The Adam and AdamW tests fail when configs (optim_input.kwargs) are {'lr': 0.01, 'fused': False}, {'lr': 0.01, 'fused': True}). This is with the AssertionError context.

When you say "with the AssertionError context", do you mean with the context enabled? If so, I am not surprised that these two basic configs now pass + fail the assertion error. Or do you mean that these configs are what causes the major errors?

Ah, is that the only SGD config failing? If you comment it out, does the rest of the test pass?

And are these all for CPU?

Reply:
Sorry, I wasn't clear about the failing cases. 

1. When the `context` is `partial(self.assertRaises, AssertionError)` for Adam and AdamW, the tests fail for configs `{'lr': 0.01, 'fused': False}`, `{'lr': 0.01, 'fused': True}` with the error `AssertionError: AssertionError not raised`. 
2. When the `context` is `contextlib.nullcontext`, the tests fail for all the configs with the error `AssertionError: Tensor-likes are not close!`. In this case, I am confused as to why is the error being thrown even for the configs where case 1 fails, the mismatch elements percentage is around 3.1% for `{'lr': 0.01, 'fused': False}`, `{'lr': 0.01, 'fused': True}` but either 39.1% or 100% for other configs.

For SGD, that is the only config that fails. I am able to pass the test if I skip this config.

These all tests are for CPU. 
I see that the Native devices are ('cpu', 'cuda', 'meta'). I understand cuda tests are skipped since my machine does not have cuda support. What about meta device type?

Reply:
Okay, debugging these errors will likely be a more involved task. I propose:
- opening an issue tracking your observations here about this particular test (what fails, what doesn't when the context is on/off).
- no longer using the _get_optim_inputs_including_global_cliquey_kwargs to get all the kwargs, but just having the kwargs  be parametrized like in https://github.com/pytorch/pytorch/blob/main/test/test_optim.py#L689 but with the impl being "forloop", "foreach", "fused". 
   - Depending on the impl, you can make the kwargs {"foreach": False}, {"foreach": True}, and {"fused": True} respectively. 
- This way we consolidate into one test but avoid signing up to address all the issues now.

And to answer your q about the meta device, it does input and output shape checking, as the tensors are not actually backed by real storage!

Reply:
I can pass the cpu tests locally by skipping the failing configs from the `_get_optim_inputs_including_global_cliquey_kwargs`. 

Should I try pushing it and checking if the CI passes?

Reply:
I am suggesting not using _get_optim_inputs_including_global_cliquey_kwargs at all, but to just use 

kwargs = {"foreach": False} if forloop else ...
optim_cls(params, **kwargs)

Once you push that change, the tests should pass since it is not adding more coverage compared to before.

Reply:
But yes, please file the issue!